drivers/counter/counter_ifx_cat1.c

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright (c) 2023 Cypress Semiconductor Corporation (an Infineon company) or
+ * an affiliate of Cypress Semiconductor Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Counter driver for Infineon CAT1 MCU family.
+ */
+
+#define DT_DRV_COMPAT infineon_cat1_counter
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <cyhal_timer.h>
+#include <cyhal_gpio.h>
+
+#include <cyhal_tcpwm_common.h>
+#include <zephyr/dt-bindings/pinctrl/ifx_cat1-pinctrl.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ifx_cat1_counter, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_counter_config {
+	struct counter_config_info counter_info;
+	TCPWM_CNT_Type *reg_addr;
+	cyhal_gpio_t external_pin;
+	IRQn_Type irqn;
+	uint8_t irq_priority;
+};
+
+/* Data structure */
+struct ifx_cat1_counter_data {
+	cyhal_timer_t counter_obj;
+	cyhal_timer_cfg_t counter_cfg;
+	struct counter_alarm_cfg alarm_cfg_counter;
+	struct counter_top_cfg top_value_cfg_counter;
+	uint32_t guard_period;
+	cyhal_resource_inst_t hw_resource;
+	cyhal_source_t signal_source;
+	bool alarm_irq_flag;
+};
+
+/* Default Counter configuration structure */
+static const cy_stc_tcpwm_counter_config_t cyhal_timer_default_config = {
+	.period = 32768,
+	.clockPrescaler = CY_TCPWM_COUNTER_PRESCALER_DIVBY_1,
+	.runMode = CY_TCPWM_COUNTER_CONTINUOUS,
+	.countDirection = CY_TCPWM_COUNTER_COUNT_UP,
+	.compareOrCapture = CY_TCPWM_COUNTER_MODE_CAPTURE,
+	.compare0 = 16384,
+	.compare1 = 16384,
+	.enableCompareSwap = false,
+	.interruptSources = CY_TCPWM_INT_NONE,
+	.captureInputMode = 0x3U,
+	.captureInput = CY_TCPWM_INPUT_0,
+	.reloadInputMode = 0x3U,
+	.reloadInput = CY_TCPWM_INPUT_0,
+	.startInputMode = 0x3U,
+	.startInput = CY_TCPWM_INPUT_0,
+	.stopInputMode = 0x3U,
+	.stopInput = CY_TCPWM_INPUT_0,
+	.countInputMode = 0x3U,
+	.countInput = CY_TCPWM_INPUT_1,
+};
+
+static int get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
+{
+	int ret = -EINVAL;
+	uint32_t i;
+
+	for (i = 0u; i < _CYHAL_TCPWM_INSTANCES; i++) {
+		uintptr_t base = POINTER_TO_UINT(_CYHAL_TCPWM_DATA[i].base);
+		uintptr_t cnt = POINTER_TO_UINT(_CYHAL_TCPWM_DATA[i].base->CNT);
+		uintptr_t reg_addr_ptr = POINTER_TO_UINT(reg_addr);
+		uintptr_t end_addr = base + sizeof(TCPWM_Type);
+
+		if ((reg_addr_ptr > base) && (reg_addr_ptr < end_addr)) {
+
+			hw_resource->type = CYHAL_RSC_TCPWM;
+			hw_resource->block_num = i;
+			hw_resource->channel_num = ((reg_addr_ptr - cnt) / sizeof(TCPWM_CNT_Type));
+
+			if (hw_resource->channel_num < _CYHAL_TCPWM_DATA[i].num_channels) {
+				ret = 0;
+			}
+			break;
```

Comment:
The error path here is quite hard to follow, if I understand the intent could you change it to:
- `return 0` on like 87
- `return -EINVAL` on line 89 and 93

That should be easier to understand. Even better flip the condition on line 86 and make that the error path.

Reply:
Flipped the conditions to make it easier to understand